### PR TITLE
Adds a notification plugin, and a few tweaks to help support this

### PR DESCRIPTION
**Pull request by jghibiki**
### Please reference an issue, or fill out the below information

- Issue: 

---

* **Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

Adds a notification plugin, with back-ends for pushbullet, email, and slack (the underlying library notifiers supports additional platforms).

Also updates the Server API system to add a ServerResponse object to better standardize the responses from the plugin/builtin server api hooks.

Also refactors the IntenEvent system to have it's own builtin. Added a handler that will take an Intent object and convert it to a typed Event subclass object that can be submitted to the event bus. This allows plugins and builtins to be loosly coupled as a builtin or plugin can create an intent which will be converted to a required typed event that another builtin or plugin can handle. 

e.g. 
```
IntentEvent({"type": "msa.plugins.my_plugin.events.MyCustomEvent", "context": { <my event init data}) 
  -> IntentHandler 
    -> MyCustomEvent instance 
      -> event bus
```

For example to trigger a notification via the notification plugin via the cli (requires that the notification plugin is configured, and configured to be loaded in the msa_config.json):
```python
# Tightly coupled version, requires being able to import the SendNotificationEvent
from msa.plugins.notifications.events import SendNotificationEvent  
from msa.plugins.notifications.notification_providers import NotificationProvider
a = SendNotificationEvent()
    .init({
        "provider": NotificationProvider.pushbullet.value, 
        "title": "test!", 
        "message": "this is a sample message"
    })
    .network_propagate()

# Loosely coupled version:
from msa.builtins.intents.events.IntentEvent
a = IntentEvent()
    .init({
        "type": "msa.plugins.notifications.events.SendNotificationEvent",
        "context": {
            "provider": "pushbullet",
            "title": "test!",
            "message": "this is a sample message"
    }}
   .network_propagate()
```

---

❗️🖊   If you have not yet, please sign our [Contributor License Agreement](https://docs.google.com/forms/d/e/1FAIpQLSfJPpwIv8MJkzxj41zBhnBDkrziQr9Tn_Z8hl5fRyY0xZnDuQ/viewform). 
Your PR will not be accepted until we have confirmed that you have signed.

### CODE REVIEW

**Comment by jghibiki** [comment]
Test Coverage is up to 70%, new plugins are fully tested.

**Comment by jghibiki** [comment]
@dekarrin Any idea why the build would be failing as it is on the git clone step?

**Comment by jghibiki** [comment]
Hm, well the clone step worked this time. Must have been a temp. problem.

File: .drone.yml [hunk 1]
```diff
@@ -37,12 +37,12 @@ pipeline:
...
-    image: python:3
+    image: python:3.8
     commands:
       - cd python
-      - pip install poetry
-      - poetry install 
-      - "TEST=1 ./test_and_coverage.sh"
+      - curl -sSL https://raw.githubusercontent.com/python-poetry/poetry/master/get-poetry.py | python -
+      - $HOME/.poetry/bin/poetry install 
+      - $HOME/.poetry/bin/poetry run "TEST=1 ./test_and_coverage.sh"
```
Comment (dekarrin) [inline .drone.yml:45]: it looks like passing the whole command sequence/argument as a single string is making something further down the line treat it as a whole file with a space in its name. is it possible to move the TEST=1 to the environment section? i believe vars in the environment section will be exported so should be visible by poetry and child processes, unless im mistaken glubglub.

```suggestion
      - $HOME/.poetry/bin/poetry run ./test_and_coverage.sh
    environment:
      - TEST=1
```

**Comment by jghibiki** [comment]
:party build finally succeeded!

**Comment by jghibiki** [comment]
@dekarrin whenever you have some free time, i think this is ready for review.

**Review by dekarrin — APPROVED** [review]

**Pull request merged**